### PR TITLE
PCHR-2410: Refactor + make one api call to update document status

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
Currently, for the documents that are the part of the assignment, when updated the new revision is created and when document is updated after page reload only the second-most-recent revision is listed and when opening the document, console error is thrown due to the second-most-recent  revision being saved with `is_current_revision = 0`. But documents with `is_current_revision = 0` can't be fetched while opening document modal.

This PR  fixes the issue by limiting the api call to update the document only once as a result the revision creation is avoided and after that only the latest revision will be listed in SSP documents list.

## Before
- When the document (as a part of assignment) is updated from SSP, console error is displayed as new revision is created on opening updated document.
before-iloveimg-compressed

## After
- When the document (as a part of assignment) is updated from SSP, console error is not displayed as new revision is not created on opening updated document.
tfinal-iloveimg-compressed

## Technical Details
1. Previously, on saving or updating the document (as part of assignment), an extra request is made to update the document status. This created an new revision with `is_current_revision = 0`. Since in T&A only documents having `is_current_revision=1` are only  fetched, it raises the console error that the document is not available. To fix this issue following steps have been taken:

1.1. Removed the calls to make the second request that updated the document
```js
if (result.files.length && !!result.files[0].result) {
   DocumentService.save({
       id: result.document.id,
       status_id: '1' // 1 => 'awaiting upload'
    }).then(function (results) {
       vm.document.status_id = results.status_id;
       $modalInstance.close(vm.document); 
    });
} else if (result.files.length && !!result.files[0].values[0].result) {
    DocumentService.save({
       id: result.document.id,
       status_id: vm.isRole('admin') ? '3' : '1' // 3 => 'approved', 1 => 'awaiting upload'
    }).then(function (results) {
       vm.document.status_id = results.status_id;
       $modalInstance.close(vm.document);
    });
} else {
   $modalInstance.close(vm.document);
}
```
1.2. The status of the document is determined before saving/updating the document and saving the document with the finalized status. 
```js
var documentStatus= vm.document.status_id  ;
var file;
var promiseFilesDelete = [];
  
$scope.$broadcast('ct-spinner-show');
  
documentStatus = vm.isRole('admin') ? '3' : (vm.files.length || vm.uploader.queue.length ? '2' : '1');
...
doc.status_id = documentStatus;
```
1.3 Remove the unnecessary variables `var uploader` and `var filesTrash` and use their respective `vm.uploader` and `vm.filesTrash`.

## Comments
- The file should be made required, which will be fixed in the next PR.

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 1]
```diff
@@ -318,6 +318,122 @@ define([
           expect(DocumentService.save).toHaveBeenCalledWith(jasmine.objectContaining({ activity_date_time: '' }));
         });
       });
+
+      describe('document contains files', function () {
+        beforeEach(function () {
+          angular.extend(controller.document, mockDocument);
+        });
+
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: Inconsistent spacing between `beforeEach` and `describe` blocks. Please fix it in all places in the document.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 2]
```diff
@@ -318,6 +318,122 @@ define([
           expect(DocumentService.save).toHaveBeenCalledWith(jasmine.objectContaining({ activity_date_time: '' }));
         });
       });
+
+      describe('document contains files', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: I am trying to read tests and I am not quite sure I understand the difference between `document contains files` and `document already has attachments`. Could you clarify please?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 3]
```diff
@@ -318,6 +318,122 @@ define([
...
+          beforeEach(function () {
+            controller.uploader.queue.length = 3;
+          });
+          describe('user is admin', function () {
+            beforeEach(function () {
+              controller.role = 'admin';
+              controller.confirm();
+            });
+
+            it('sets document status to be approved', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: `to Be Approved` or `to Approved` or `to To Be Approved`?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 4]
```diff
@@ -318,6 +318,122 @@ define([
...
+          beforeEach(function () {
+            controller.files.length = 2;
+          });
+          describe('user is admin', function () {
+            beforeEach(function () {
+              controller.role = 'admin';
+              controller.confirm();
+            });
+
+            it('sets the document status to approved', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: `sets document status to Approved` or `sets document status to "Approved"`, please select one of the suggested or propose your version. The key is to show that "Approved" is a status not just a word. Please fix this in other places in the test.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 5]
```diff
@@ -318,6 +318,122 @@ define([
...
+            });
+
+            it('sets document status to be awaiting approval', function () {
+              expect(DocumentService.save).toHaveBeenCalledWith(jasmine.objectContaining({ status_id: '2' }));
+            });
+          });
+        });
+      });
+
+      describe("document doesn\'t contains files", function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: doesn\'t contain

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js [hunk 6]
```diff
@@ -318,6 +318,122 @@ define([
...
+        beforeEach(function () {
+          angular.extend(controller.document, mockDocument);
+        });
+
+        describe('document does not have any attachment', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: any attachments

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js [hunk 7]
```diff
@@ -181,18 +181,25 @@ define([
           return;
         }
 
-        var uploader = vm.uploader;
+        var documentStatus= vm.document.status_id  ;
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js [hunk 8]
```diff
@@ -227,31 +234,13 @@ define([
...
-          }
-
-          vm.document.id = result.document.id;
           vm.document.case_id = result.document.case_id;
           vm.document.file_count = vm.files.length + uploader.queue.length;
+          vm.document.id = result.document.id;
+          vm.document.status_id = result.status_id;
 
           AssignmentService.updateTab();
+          $modalInstance.close(vm.document);
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js [hunk 9]
```diff
@@ -181,18 +181,25 @@ define([
...
+        var documentStatus= vm.document.status_id  ;
+        var file;
         var filesTrash = vm.filesTrash;
         var promiseFilesDelete = [];
-        var file;
+        var uploader = vm.uploader;
 
         $scope.$broadcast('ct-spinner-show');
 
+        if (vm.files.length || vm.uploader.queue.length) {
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.js]: No particular reason, should be used `vm.uploader`.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js (hunk 2 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: 1. `document contains files` should be `document has attachments in upload queue` 
2. `document already has attachments` should be `document has existing attachments`
I will update it in test cases.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js (hunk 1 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js (hunk 4 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js (hunk 3 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: For me `to Approved` looks good. Will update the test cases.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js (hunk 6 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.spec.js]: done!

**Comment by bmnepali** [comment]
Fixed merge conflicts by rebase.

**Pull request merged**